_releases/2022-04-06-1.4.0-released.md

```diff
@@ -0,0 +1,80 @@
+---
+title: Crystal 1.4.0 is released!
+version: 1.4.0
+summary:
+thumbnail: +
+author:
+---
+
+We are delivering a new release with several bugfixes and improvements. Below we list the most important or interesting changes, without mentioning several bugfixes and smaller enhancements. For more details, visit the [changelog](https://github.com/crystal-lang/crystal/releases/tag/1.4.0). Breaking changes are marked with ⚠️.
+
+Pre-built packages are available on [GitHub Releases](https://github.com/crystal-lang/crystal/releases/tag/1.4.0) and our official distribution channels.
+See [crystal-lang.org/install](https://crystal-lang.org/install/) for installation instructions.
+
+## Stats
+
+In this release we included [132 changes since the 1.3.2 release](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+milestone%3A1.4.0) by ?? contributors. We thank all the effort put into improving the language! ❤️
+
+## Towards WASM support
+
+Version 1.4.0 ships with, at the moment, minimal support for compiling into WebAssembly, linking it with a WASI-based LibC. There are several missing important pieces, check the [PR](https://github.com/crystal-lang/crystal/pull/10870). The updated instructions to try it out are:
+
+1. Write a simple Crystal program, let's say: `puts "Hello WebAssembly!"` and name it `main.cr`.
+
+2. Build the WebAssembly module:
+
+```bash
+$ crystal build main.cr --cross-compile --target wasm32-unknown-wasi
+```
+
+This will generate a `main.wasm` file that is actually a binary wasm file.
+
+3. For the program to work you need a LibC based on WASI. You can download a precompiled one from [wasi-sdk](https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-14/wasi-sysroot-14.0.tar.gz).
+
+4. Additionally, you will need [compiler-rt](https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-14/libclang_rt.builtins-wasm32-wasi-14.0.tar.gz) and `libpcre`, which @lguillaume compiled to wasi [here](https://github.com/lbguilherme/crystal/files/7791111/libpcre-8.45.tar.gz).
+
+5. From the files above, place `libc.a`, `libpcre.a`, `libclang_rt.builtins-wasm32.a`, and `librt.a` in a folder of your choice, say `./lib/`.
+
+6. Link your program and these libraries into a final WebAssembly module:
+
+```bash
+$ wasm-ld main.wasm -o main-final.wasm -L./lib -lc -lpcre -lrt -lclang_rt.builtins-wasm32
+```
+
+5. Now you have a wasm file that is a final module that can run on any WASI-capable runtime, like Wasmer.
+
+```bash
+$ wasmer main-final.wasm 
+Hello WebAssembly!
+```
```

Comment:
If https://github.com/crystal-lang/crystal/pull/11948 gets merged, then the instructions are:

```md
1. Write a simple Crystal program, let's say: `puts "Hello WebAssembly!"` and name it `main.cr`.
You can also try more complicated programs, but keep in mind that this is still experimental
and not ready for production.
 
2. You need to have `wasm-ld` installed on your system, it is the WebAssembly linker from LLVM.
On some systems, it comes by installing `lld`. Confirm that it is installed by using
`wasm-ld --version`. Please also check that its version is similar to Crystal's LLVM version
(`crystal --version`).

3. You will need `libc` and `libpcre` compiled to the `wasm32-wasi` target. You can fetch a
precompiled version of them here: https://github.com/lbguilherme/wasm-libs/releases.
Check for the `wasm32-wasi-libs.tar.gz` asset, download it and extract it.

4. Compile it with
`CRYSTAL_LIBRARY_PATH=$PWD/wasm32-wasi-libs crystal build main.cr -o main.wasm --target wasm32-wasi`.

5. Run the WebAssembly module with `wasmer main.wasm` or `wasmtime main.wasm` and have fun.
```

If it is not, then step 4 becomes:

```md
4. Cross-compile it with `crystal build main.cr --cross-compile --target wasm32-wasi`.

5. Link it with `wasm-ld main.wasm -o main-final.wasm  -L "$PWD/wasm32-wasi-libs" -lc -lpcre`
```